### PR TITLE
type: encapsulate Time API

### PR DESCRIPTION
**Pull request by TennyZhuang**
### What problem does this PR solve? 

This is a prepare PR to #14278

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

Related changes

Release note

 - encapsulate Time API

### CODE REVIEW

File: distsql/distsql.go [hunk 1]
```diff
@@ -166,8 +166,8 @@ func canUseChunkRPC(ctx sessionctx.Context) bool {
 	return true
 }
 
-var supportedAlignment = !(unsafe.Sizeof(types.MysqlTime{}) != 16 ||
```
Comment (breeswish) [inline distsql/distsql.go]: Why is this changed?

Comment (TennyZhuang) [reply distsql/distsql.go]: will fix it

File: server/driver_tidb.go [hunk 2]
```diff
@@ -480,7 +480,7 @@ func convertColumnInfo(fld *ast.ResultField) (ci *ColumnInfo) {
 	// Keep things compatible for old clients.
 	// Refer to mysql-server/sql/protocol.cc send_result_set_metadata()
 	if ci.Type == mysql.TypeVarchar {
-		ci.Type = mysql.TypeVarString
+		ci.SetType(mysql.TypeVarString)
```
Comment (breeswish) [inline server/driver_tidb.go]: ???

File: types/convert.go [hunk 3]
```diff
@@ -341,12 +340,10 @@ func NumberToDuration(number int64, fsp int8) (Duration, error) {
 				return dur, errors.Trace(err1)
 			}
 		}
-		dur, err1 := MaxMySQLTime(fsp).ConvertToDuration()
-		terror.Log(err1)
+		dur := MaxMySQLTime(fsp)
```
Comment (breeswish) [inline types/convert.go]: why this is ok?

File: types/datum.go [hunk 4]
```diff
@@ -1051,7 +1051,7 @@ func (d *Datum) convertToMysqlTime(sc *stmtctx.StatementContext, target *FieldTy
 	}
 	if tp == mysql.TypeDate {
 		// Truncate hh:mm:ss part if the type is Date.
-		t.Time = FromDate(t.Time.Year(), t.Time.Month(), t.Time.Day(), 0, 0, 0, 0)
+		t = NewTime(FromDate(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0), t.Type(), t.Fsp())
```
Comment (breeswish) [inline types/datum.go]: t.SetDateTimePart?

File: types/time_test.go [hunk 5]
```diff
@@ -832,7 +825,8 @@ func (s *testTimeSuite) TestConvert(c *C) {
 		n := time.Date(year, month, day, 0, 0, 0, 0, sc.TimeZone)
 		t, err := v.ConvertToTime(sc, mysql.TypeDatetime)
 		c.Assert(err, IsNil)
-		t1, _ := t.Time.GoTime(sc.TimeZone)
+		// TODO: Consider time_zone variable.
```
Comment (breeswish) [inline types/time_test.go]: todo?

File: types/time_test.go [hunk 6]
```diff
@@ -997,10 +983,9 @@ func (s *testTimeSuite) TestConvertTimeZone(c *C) {
 	}
 
 	for _, test := range tests {
-		var t types.Time
-		t.Time = test.input
+		t := types.NewTime(test.input, mysql.TypeDatetime, 0)
```
Comment (breeswish) [inline types/time_test.go]: mysql.TypeDatetime -> 0?

File: types/time_test.go [hunk 7]
```diff
@@ -997,10 +983,9 @@ func (s *testTimeSuite) TestConvertTimeZone(c *C) {
...
 		t.ConvertTimeZone(test.from, test.to)
-		c.Assert(t.Compare(types.Time{Time: test.expect}), Equals, 0)
+		c.Assert(t.Compare(types.NewTime(test.expect, mysql.TypeDatetime, 0)), Equals, 0)
```
Comment (breeswish) [inline types/time_test.go]: mysql.TypeDatetime -> 0?

File: types/time_test.go [hunk 8]
```diff
@@ -1030,7 +1015,7 @@ func (s *testTimeSuite) TestTimeAdd(c *C) {
 		c.Assert(err, IsNil)
 		v2, err := v1.Add(sc, dur)
 		c.Assert(err, IsNil)
-		c.Assert(v2.Compare(result), Equals, 0, Commentf("%v %v", v2.Time, result.Time))
+		c.Assert(v2.Compare(result), Equals, 0, Commentf("%v %v", v2, result))
```
Comment (breeswish) [inline types/time_test.go]: v2.GetDateTimePart, result.GetDateTimePart?

File: types/time_test.go [hunk 9]
```diff
@@ -1313,21 +1298,17 @@ func (s *testTimeSuite) TestExtractDurationValue(c *C) {
 
 func (s *testTimeSuite) TestCurrentTime(c *C) {
 	res := types.CurrentTime(mysql.TypeTimestamp)
-	c.Assert(res.Time, NotNil)
-	c.Assert(res.Type, Equals, mysql.TypeTimestamp)
-	c.Assert(res.Fsp, Equals, int8(0))
+	c.Assert(res.Type(), Equals, mysql.TypeTimestamp)
```
Comment (breeswish) [inline types/time_test.go]: assert res.GetDateTimePart() > 0?

File: types/time_test.go [hunk 10]
```diff
@@ -1313,21 +1298,17 @@ func (s *testTimeSuite) TestExtractDurationValue(c *C) {
...
-	in := types.Time{
-		Time: types.ZeroTime,
-		Type: mysql.TypeTimestamp,
-		Fsp:  types.DefaultFsp,
-	}
+	var in types.Time
+	in = types.NewTime(types.ZeroTime, mysql.TypeTimestamp, types.DefaultFsp)
 	c.Assert(in.InvalidZero(), Equals, true)
-	in.Time = types.FromDate(2019, 00, 00, 00, 00, 00, 00)
+	in = types.NewTime(types.FromDate(2019, 00, 00, 00, 00, 00, 00), mysql.TypeTimestamp, types.DefaultFsp)
```
Comment (breeswish) [inline types/time_test.go]: mysql.TypeTimestamp -> 0?

File: types/time_test.go [hunk 11]
```diff
@@ -1313,21 +1298,17 @@ func (s *testTimeSuite) TestExtractDurationValue(c *C) {
...
 	c.Assert(in.InvalidZero(), Equals, true)
-	in.Time = types.FromDate(2019, 04, 12, 12, 00, 00, 00)
+	in = types.NewTime(types.FromDate(2019, 04, 12, 12, 00, 00, 00), mysql.TypeTimestamp, types.DefaultFsp)
```
Comment (breeswish) [inline types/time_test.go]: in.SetDateTimePart?

File: expression/builtin_cast_vec_test.go [hunk 12]
```diff
@@ -136,10 +137,11 @@ func (g *datetimeJSONGener) gen() interface{} {
 	minute := rand.Intn(60)
 	second := rand.Intn(60)
 	microsecond := rand.Intn(1000000)
-	d := types.Time{
-		Time: types.FromDate(year, month, day, hour, minute, second, microsecond),
-		Fsp:  3,
-	}
+	d := types.NewTime(
+		types.FromDate(year, month, day, hour, minute, second, microsecond),
+		mysql.TypeDatetime,
```
Comment (breeswish) [inline expression/builtin_cast_vec_test.go]: 0?

File: types/convert_test.go [hunk 13]
```diff
@@ -915,8 +912,8 @@ func testConvertTimeTimeZone(c *C, sc *stmtctx.StatementContext) {
 		nd, err := d.ConvertTo(sc, test.target)
 		c.Assert(err, IsNil)
 		t := nd.GetMysqlTime()
-		c.Assert(t.Type, Equals, test.expect.Type)
-		c.Assert(t.Time, Equals, test.expect.Time)
+		c.Assert(t.Type(), Equals, test.expect.Type)
```
Comment (breeswish) [inline types/convert_test.go]: `test.expect.Type()`?

File: types/time_test.go [hunk 14]
```diff
@@ -453,22 +453,18 @@ func (s *testTimeSuite) TestCodec(c *C) {
...
-	var t1 types.Time
-	t1.Type = mysql.TypeTimestamp
-	t1.Time = types.FromGoTime(time.Now())
+	t1 := types.NewTime(types.FromGoTime(time.Now()), mysql.TypeTimestamp, 0)
 	packed, err := t1.ToPackedUint()
 	c.Assert(err, IsNil)
 
-	var t2 types.Time
-	t2.Type = mysql.TypeTimestamp
+	t2 := types.NewTime(types.ZeroTime, mysql.TypeTimestamp, 0)
```
Comment (breeswish) [inline types/time_test.go]: call it `types.ZeroCoreTime`?

File: expression/constant_test.go [hunk 15]
```diff
@@ -483,7 +480,7 @@ func (*testExpressionSuite) TestDeferredExprNotNull(c *C) {
 	xDec, _, _ := cst.EvalDecimal(ctx, chunk.Row{})
 	c.Assert(xDec.Compare(m.i.(*types.MyDecimal)), Equals, 0)
 
-	m.i = types.Time{}
+	m.i = types.NewTimeZeroValue()
```
Comment (breeswish) [inline expression/constant_test.go]: How about:
`types.ZeroTime` -> `types.ZeroCoreTime`
`types.NewTimeZeroValue()` -> `types.ZeroTime`?

File: expression/builtin_cast_vec_test.go (hunk 12 above)
Comment (TennyZhuang) [reply expression/builtin_cast_vec_test.go]: guess datetime is better from the context.

**Comment by breeswish** [comment]
LGTM

**Comment by breeswish** [comment]
LGTM

File: expression/builtin_cast_vec_test.go [hunk 16]
```diff
@@ -136,10 +136,11 @@ func (g *datetimeJSONGener) gen() interface{} {
 	minute := rand.Intn(60)
 	second := rand.Intn(60)
 	microsecond := rand.Intn(1000000)
-	d := types.Time{
-		Time: types.FromDate(year, month, day, hour, minute, second, microsecond),
-		Fsp:  3,
-	}
+	d := types.NewTime(
+		types.FromDate(year, month, day, hour, minute, second, microsecond),
+		0,
```
Comment (lonng) [inline expression/builtin_cast_vec_test.go]: `0` indicates it is a `Decimal`

File: types/time.go [hunk 17]
```diff
@@ -216,47 +219,147 @@ func FromDate(year int, month int, day int, hour int, minute int, second int, mi
...
+
 // Clock returns the hour, minute, and second within the day specified by t.
 func (t Time) Clock() (hour int, minute int, second int) {
-	return t.Time.Hour(), t.Time.Minute(), t.Time.Second()
+	return t.time.Hour(), t.time.Minute(), t.time.Second()
 }
 
 // Time is the struct for handling datetime, timestamp and date.
-// TODO: check if need a NewTime function to set Fsp default value?
+// TODO: check if need a NewTime function to set.fsp default value?
```
Comment (lonng) [inline types/time.go]: Accidentally injured?

File: types/time.go [hunk 18]
```diff
@@ -216,47 +219,147 @@ func FromDate(year int, month int, day int, hour int, minute int, second int, mi
...
 	// Fsp is short for Fractional Seconds Precision.
 	// See http://dev.mysql.com/doc/refman/5.7/en/fractional-seconds.html
-	Fsp int8
+	fsp int8
 }
 
-// MaxMySQLTime returns Time with maximum mysql time type.
-func MaxMySQLTime(fsp int8) Time {
-	return Time{Time: FromDate(0, 0, 0, TimeMaxHour, TimeMaxMinute, TimeMaxSecond, 0), Type: mysql.TypeDuration, Fsp: fsp}
+// NewTime construct time from core time, type and fsp.
```
Comment (lonng) [inline types/time.go]:
```suggestion
// NewTime constructs time from core time, type and fsp.
```

File: types/time.go [hunk 19]
```diff
@@ -216,47 +219,147 @@ func FromDate(year int, month int, day int, hour int, minute int, second int, mi
...
+func NewTime(mt MysqlTime, tp uint8, fsp int8) Time {
+	return Time{
+		time: mt,
+		tp:   tp,
+		fsp:  fsp,
+	}
+}
+
+// Year return year value.
```
Comment (lonng) [inline types/time.go]:
```suggestion
// Year returns year value.
```

File: types/time.go [hunk 20]
```diff
@@ -216,47 +219,147 @@ func FromDate(year int, month int, day int, hour int, minute int, second int, mi
...
+
+// Second return second value.
+func (t Time) Second() int {
+	return t.time.Second()
+}
+
+// Microsecond return microsecond value.
+func (t Time) Microsecond() int {
+	return t.time.Microsecond()
+}
```
Comment (lonng) [inline types/time.go]: All comments should be refined.
